components/server/src/workspace/workspace-starter.ts

```diff
@@ -332,46 +338,123 @@ export class WorkspaceStarter {
+                    this.reconcileWorkspaceStart(ctx, instance.id);
+                    return;
+
+                case "pending":
+                    // !!! Note: during pending, the handover between app-cluster and ws-manager happens. !!!
+                    // This means:
+                    //  - there is a control loop on ws-manager-bridge that checks for an upper limit a instance may be in pending phase
+                    //  - it will take some time after calling ws-manager to see the first status update
+                    if (durationLongerThanSeconds(Date.parse(instance.creationTime), 30)) {
+                        // In 99.9% this is due to timing, so we need to be a bit cautious here not to spam ourselves.
+                        return;
+                    }
+
+                    // Our time has come!
+                    this.reconcileWorkspaceStart(ctx, instance.id);
+                    return;
+
+                default:
+                    return;
             }
-            span.setTag("needsImageBuild", needsImageBuild);
-        } catch (err) {
-            // if we fail to check if the workspace needs an image build (e.g. becuase the image builder is unavailable),
-            // we must properly fail the workspace instance, i.e. set its status to stopped, deal with prebuilds etc.
-            //
-            // Once we've reached actuallyStartWorkspace that function will take care of failing the instance.
-            await this.failInstanceStart(ctx, err, workspace, instance);
-            throw err;
         }
+    }
 
-        if (needsImageBuild) {
-            this.actuallyStartWorkspace({ span }, instance, workspace, user, envVars, forceRebuild).catch((err) =>
-                log.error("actuallyStartWorkspace", err),
-            );
-            return { instanceID: instance.id };
-        }
+    public reconcileWorkspaceStart(_ctx: TraceContext, instanceId: string) {
+        const ctx = TraceContext.childContext("reconcileWorkspaceStart", _ctx);
+
+        // We try to acquire a mutex here, which we intend to hold until the workspace start request is sent to ws-manager.
+        // In case this container dies for whatever reason, the mutex is eventually released, and the instance can be picked up
+        // by another server process (cmp. controlAllOurStartingWorkspaces() above).
+        this.redisMutex
+            .using(
+                ["workspace-instance-start-" + instanceId],
+                5000, // after 5s without extension the lock is released
+                async (abortSignal) => {
+                    const instance = await this.workspaceDb.trace({}).findInstanceById(instanceId);
+                    if (!instance) {
+                        ctx.span.finish();
+                        throw new Error("cannot find workspace for instance");
+                    }
+                    const workspace = await this.workspaceDb.trace(ctx).findById(instance.workspaceId);
```

Comment:
we could check if the instance is not already in a phase (e.g. `running`) where this method has nothing to do and return early.

Reply:
Good point! This should be the primary point for checking that anyway. :100:

Reply:
Yes, the checks you do in the reconile loop could be done here.